File: MergeSort.java
import java.util.Random;

public class MergeSort {

    private static void merge(int[] Arr , int[] leftHalf , int[] rightHalf ){
        int leftSize = leftHalf.length;
        int rightSize = rightHalf.length;
        int i = 0 , j= 0 , k = 0;

        while(i < leftSize && j < rightSize){
            if(leftHalf[i] <= rightHalf[j]){
                Arr[k] = leftHalf[i];
                i++;
            }else {
                Arr[k] = rightHalf[j];
                j++;
            }
            k++;
        }

        while(i < leftSize){
            Arr[k] = leftHalf[i];
            i++;
            k++;
        }

        while(j < rightSize){
            Arr[k] = rightHalf[j];
            j++;
            k++;
        }

    }

    private static void mergeSort(int[] Arr){
        int n = Arr.length;

        if(n < 2){
            return;
        }

        int mid = n/2;
        int[] left = new int[mid];
        int[] right = new int[n - mid];

        for (int i = 0; i < mid; i++) {
            left[i] = Arr[i];
        }

        for (int i = mid; i < n ; i++) {
            right[i  - mid] = Arr[i];
        }

        mergeSort(left);
        mergeSort(right);

        //Merge
        merge(Arr , left , right);

    }


    public static void main(String[] args) {
        Random rand = new Random();
        int[] arr = new int[100000];
        int n = arr.length;

        for (int i = 0; i < arr.length; i++) {
            arr[i] = rand.nextInt(1000000);
        }

        mergeSort(arr);

        for (int i = 0 ; i < n ; i++){
            System.out.println(arr[i] + " ");
        }
        System.out.println();
    }
}

File: MergeSort.cpp
#include <bits/stdc++.h>
using namespace std;

void merge(vector<int> &arr , int low , int mid , int high){
    vector <int> temp; // Creating a temporary array
    int left = low;
    int right = mid+1;

    while (left <= mid && right <= high)
    {
        if(arr[left] <= arr[right]){
            temp.push_back(arr[left]);
            left++;
        }else{
            temp.push_back(arr[right]);
            right++;
        }
    }


    //Remaining elements from the left part

    while(left <= mid){
        temp.push_back(arr[left]);
        left++;
    }

    //Remaining elements of the right part

    while (right <= high )
    {
        temp.push_back(arr[right]);
        right++;
    }

    //copying elements from temp to main array
    
    for (int i = low; i <= high; i++)
    {
        /* code */

        arr[i] = temp[i-low];
    }
      
}

void mergeSort(vector <int> &arr , int low , int high){
    if(low >= high) return;

    int mid = (high + low)/2;
    mergeSort(arr , low , mid);
    mergeSort(arr , mid+1 , high);
    merge(arr , low , mid , high);
}



int main(){

    vector <int> A = {1 , 6 , 3 , 6 , 8 , 10 , 0 , 4 , 9};
    
    mergeSort(A , 0 , A.size() -1);

    for(int i = 0 ; i < A.size() ; i++){
        cout << A[i] << " ";
    }

    cout << endl;

    return 0;
}